src/CodeGen_GPU_Host.cpp

```diff
@@ -511,6 +511,45 @@ Value *CodeGen_GPU_Host<CodeGen_CPU>::get_module_state(const std::string &api_un
     return module_state;
 }
 
+template<typename CodeGen_CPU>
+void CodeGen_GPU_Host<CodeGen_CPU>::visit(const Call *op) {
+    CodeGen_CPU::visit(op);
+    if (op->name == "halide_device_malloc" || op->name == "halide_copy_to_device") {
+        // Register a destructor for this buffer if this is the first
+        // device_malloc or copy_to_device for it.
+        internal_assert(op->args.size() == 2);
+        const Variable *buf_var = op->args[0].as<Variable>();
+        internal_assert(buf_var);
+        const string &buf_name = buf_var->name;
+        string destructor_name = buf_name + "_destructor";
+
+        if (!sym_exists(destructor_name)) {
```

Comment:
Maybe this is a bit premature because it looks like this is currently the only usage, but this if block looks like it would be nice to have as a CodeGen_LLVM helper function.